samza-api/src/main/java/org/apache/samza/operators/descriptors/base/system/TransformingInputDescriptorProvider.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.descriptors.base.system;
+
+import org.apache.samza.operators.descriptors.base.stream.InputDescriptor;
+import org.apache.samza.serializers.Serde;
+
+/**
+ * Interface for advanced {@code SystemDescriptors} that constrain the type of returned {@code InputDescriptors} to
+ * their own {@code InputTransformer} or {@code StreamExpander} function result types.
+ *
+ * @param <SystemMessageType> type of the system level {@code InputTransformer} or {@code StreamExpander} results
+ */
+public interface TransformingInputDescriptorProvider<SystemMessageType> {
```

Comment:
Regarding being able to use this with transformers and expanders: Do you think the term "Transforming" couples this too much with transformers and not enough with expanders? Not sure if there is another generic term that would be much better though.

Reply:
This has been discussed offline. The same implementation of StreamExpander should take different view description and generate different instance of expanded sub-DAGs. (e.g. SQLExpander will take different SQL statements and apply the same expand() method to get different sub-DAG for each SQL statement).